### PR TITLE
WIP: missing tokenBag methods

### PR DESCRIPTION
**Pull request by TheReturnOfJan**
Addresses: https://github.com/leapdao/solEVM-enforcer/issues/175

Also fixes:
https://github.com/leapdao/solEVM-enforcer/issues/177

### CODE REVIEW

**Comment by TheReturnOfJan** [comment]
NOTE: the Verifier contract tests are failing due to the bytecode becoming too large. I intend to  patch-fix this in this PR and work on this problem more generally in a separate issue.

File: contracts/EVMRuntime.sol [hunk 1]
```diff
@@ -1613,13 +1612,43 @@ contract EVMRuntime is EVMConstants {
...
-        success = state.tokenBag.transfer(
+        success = state.tokenBag.transferFrom(
           stack.target,
           state.caller,
           dest,
           amount
         );
 	returnData = abi.encodePacked(success);
+      } else if (funSig == FUNCSIG_TRANSFERFROM) {
+      	address from;
```
Comment (johannbarbie) [inline contracts/EVMRuntime.sol]:
```suggestion
        address from;
```

File: contracts/EVMRuntime.sol [hunk 2]
```diff
@@ -1613,13 +1612,43 @@ contract EVMRuntime is EVMConstants {
...
+        address to;
+      	uint amount;
+
+      	assembly {
```
Comment (johannbarbie) [inline contracts/EVMRuntime.sol]:
```suggestion
        assembly {
```

File: contracts/EVMRuntime.sol [hunk 3]
```diff
@@ -1740,7 +1769,17 @@ contract EVMRuntime is EVMConstants {
 
 	  // what happens here if we enter token intercepts instead of precompiles
 	  retEvm.returnData = ret;
-	}
+	} else if (funSig == FUNCSIG_OWNEROF) {
+          uint tokenId;
+	  // 32 length + 4 funcSig = 36
+	  assembly {
```
Comment (johannbarbie) [inline contracts/EVMRuntime.sol]:
```suggestion
          assembly {
```

File: contracts/EVMTokenBag.slb [hunk 4]
```diff
@@ -20,79 +25,143 @@ library EVMTokenBag {
...
-    
-        // sender does not have enough tokens to send
-        if (source.valueOrId < value) return false;
-    
-        // find dest and/or empty
-        uint emptyId = 1337;
+        return (token, success);
+    }
+
+    function findToken2(
```
Comment (johannbarbie) [inline contracts/EVMTokenBag.slb]: document the difference to `findToken`, including `selector` param :)

File: contracts/EVMTokenBag.slb [hunk 5]
```diff
@@ -20,79 +25,143 @@ library EVMTokenBag {
...
         }
-        
-        // now do the state transition and return success
-        // OVERFLOWS???????
-        source.valueOrId -= value;
-        dest.valueOrId += value;
-        return true;
+        return (token, success);
     }
+
```
Comment (johannbarbie) [inline contracts/EVMTokenBag.slb]: why?

File: contracts/EVMTokenBag.slb (hunk 4 above)
Comment (TheReturnOfJan) [reply contracts/EVMTokenBag.slb]: Done now.

File: contracts/EVMTokenBag.slb (hunk 5 above)
Comment (TheReturnOfJan) [reply contracts/EVMTokenBag.slb]: The devil's work.

**Pull request merged**